src/dmd/dsymbolsem.d

```diff
@@ -6100,10 +6101,8 @@ Laftersemantic:
         if (tempinst.gagged)
         {
             // Errors are gagged, so remove the template instance from the
-            // instance/symbol lists we added it to and reset our state to
-            // finish clean and so we can try to instantiate it again later
+            // symbol list we added it to
             // (see https://issues.dlang.org/show_bug.cgi?id=4302 and https://issues.dlang.org/show_bug.cgi?id=6602).
-            tempdecl.removeInstance(tempdecl_instance_idx);
```

Comment:
This is the root of all evil - removing the cached instance if the errors were gagged.